### ISSUE TITLE
Bug: preserve-caught-error false positive when a local class shadows a configured error type

### ISSUE THREAD

**Issue by HakeemTheEmperor** [eslint/eslint#21285]
### Environment

Node version: v24.20.0
npm version: v11.19.0
Local ESLint version: v.10.9.1
Global ESLint version: Not found
Operating System: Windows

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
/* eslint preserve-caught-error: ["error", { errorClassNames: ["AppError"] }] */
```
</details>

```js
function makeWrapped() {
    class AppError {
        constructor(err) {
            this.original = err;
        }
    }

    try {
        doSomething();
    } catch (err) {
        throw new AppError(err);
    }
}
```

### What did you expect to happen?

**Expected behaviour**

I expected the rule to ignore the local `AppError` because it is a different class from the configured external/imported error type, and the rule should not report a missing `cause` in that case.

In other words, the rule should only enforce the configured error class when it is actually the same class being referenced in scope, not when a local declaration shadows the name.

Example of the expected outcome

```
/* eslint preserve-caught-error: ["error", { errorClassNames: ["AppError"] }] */

function makeWrapped() {
    class AppError {
        constructor(err) {
            this.original = err;
        }
    }

    try {
        doSomething();
    } catch (err) {
        throw new AppError(err);
    }
}
```

I would expect this code to be treated as valid because:

- the AppError in this scope is a local class
- it does not match the configured global/imported AppError the rule is intended to track
- the rule should not assume it is the same class just because the names match

### What actually happened?

**Actual Behaviour**

The rule reports a violation even though the class being thrown is a local, shadowing class and not the intended configured error type.

For the example below:

```
/* eslint preserve-caught-error: ["error", { errorClassNames: ["AppError"] }] */

function makeWrapped() {
    class AppError {
        constructor(err) {
            this.original = err;
        }
    }

    try {
        doSomething();
    } catch (err) {
        throw new AppError(err);
    }
}
```

the actual ESLint output is:

`error  The thrown error is not preserving the original error with a `cause` option  preserve-caught-error`

This is the false positive: the rule treats the local AppError as if it were the configured AppError, even though the name is shadowed in the current scope and the class signature is different.

**Why this is a problem**

The rule identifies configured error classes strictly by name, as noted in the docs, and does not account for local shadowing. That means it can flag a perfectly valid local wrapper class that happens to share the same name as the globally configured error class.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50XG4gIG5vLXVudXNlZC12YXJzOiBcIm9mZlwiLFxuICBuby11bmRlZjogXCJvZmZcIixcbiAgcHJlc2VydmUtY2F1Z2h0LWVycm9yOiBbXCJlcnJvclwiLCB7IGVycm9yQ2xhc3NOYW1lczogW1wiQXBwRXJyb3JcIl0gfV1cbiovXG5cbmZ1bmN0aW9uIG1ha2VXcmFwcGVkKCkge1xuICAgIGNsYXNzIEFwcEVycm9yIHtcbiAgICAgICAgY29uc3RydWN0b3IoZXJyKSB7XG4gICAgICAgICAgICB0aGlzLm9yaWdpbmFsID0gZXJyO1xuICAgICAgICB9XG4gICAgfVxuXG4gICAgdHJ5IHtcbiAgICAgICAgZG9Tb21ldGhpbmcoKTtcbiAgICB9IGNhdGNoIChlcnIpIHtcbiAgICAgICAgdGhyb3cgbmV3IEFwcEVycm9yKGVycik7XG4gICAgfVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWJpbmFyeS1leHByZXNzaW9uIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWVtcHR5LXN0YXRpYy1ibG9jayI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1uZXctbmF0aXZlLW5vbmNvbnN0cnVjdG9yIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5hc3NpZ25lZC12YXJzIjpbImVycm9yIl0sIm5vLXVuZGVmIjpbImVycm9yIl0sIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjpbImVycm9yIl0sIm5vLXVucmVhY2hhYmxlIjpbImVycm9yIl0sIm5vLXVuc2FmZS1maW5hbGx5IjpbImVycm9yIl0sIm5vLXVuc2FmZS1uZWdhdGlvbiI6WyJlcnJvciJdLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOlsiZXJyb3IiXSwibm8tdW51c2VkLWxhYmVscyI6WyJlcnJvciJdLCJuby11bnVzZWQtcHJpdmF0ZS1jbGFzcy1tZW1iZXJzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtYXNzaWdubWVudCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1jYXRjaCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWVzY2FwZSI6WyJlcnJvciJdLCJuby13aXRoIjpbImVycm9yIl0sInByZXNlcnZlLWNhdWdodC1lcnJvciI6WyJlcnJvciJdLCJyZXF1aXJlLXlpZWxkIjpbImVycm9yIl0sInVzZS1pc25hbiI6WyJlcnJvciJdLCJ2YWxpZC10eXBlb2YiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

This appears to be a false positive caused by scope shadowing. The rule matches configured error class names by identifier name alone and does not appear to account for local shadowing, so a local `AppError` class is incorrectly treated as the configured error type, even though it is a different symbol in scope.

**Comment by DMartens** [comment]
This is a [known limitation](https://eslint.org/docs/latest/rules/preserve-caught-error#known-limitations) of the rule.
We could solve this by adding another property to the object form of `errorClassNames` but I do not think this has application for real-world code as classes are usually defined in a separate file.

**Comment by HakeemTheEmperor** [comment]
Thanks for the clarification. I understand this is a known limitation and that the rule is intentionally designed as a lightweight name-based check.

I agree that a new option may not be justified for this edge case. I’d still suggest that the docs could make this limitation clearer, and I’d be happy to help with a docs improvement if that would be useful.

**Comment by DMartens** [comment]
Documentation updates which make clear things up are always welcome.
